### ISSUE TITLE
ENH: markov_switching switching needs to be in log scale

### ISSUE THREAD

**Issue by rdturnermtl** [statsmodels/statsmodels#4125]
The filters and smoothers in statsmodels.tsa.regime_switching.markov_switching (e.g., py_hamilton_filter) should be in logscale and use (scipy.misc.logsumexp etc).

The current filters (e.g. Lines 187--201) are all done in linear scale.  This can results in NaNs from underflow when a long sequence of observations with obvious states are encountered.  The standard robust implementation of HMMs is done in logscale.  See for example:
http://bozeman.genome.washington.edu/compbio/mbt599_2006/hmm_scaling_revised.pdf

**Comment by ChadFulton** [comment]
This is a good idea, and Valera took this approach when he wrote his models. It should be pretty straightforward to modify the code to do this (although the relevant code is the Cython version, not the Python version).

A PR would be very welcome of course!

**Comment by rdturnermtl** [comment]
Ok, I am more likely to break the code than the prior authors/maintainers if I re-implement it in log scale.  So, it would be best if one of them can do this.

**Comment by ChadFulton** [comment]
> Ok, I am more likely to break the code than the prior authors/maintainers if I re-implement it in log scale. So, it would be best if one of them can do this.

The main constraint for us is manpower, and so we really appreciate any additional help. I am the author / maintainer of this model, and I would be happy to help with a PR if you wanted to take a run at it.

As a side-note, were you running into underflow problems in a particular application of the model?

**Comment by wmay** [comment]
Me too, and I really need it to work, so I guess the task now falls on me. :)
I can at least do this for the python version, anyway.

For cython, is it going to defeat the purpose of cython if I call scipy's logsumexp?
Seems like the only awkward part is adding a cython logsumexp function and getting it to work with the fancy templating.

**Comment by ChadFulton** [comment]
> Me too, and I really need it to work, so I guess the task now falls on me. :)

Just to be sure, you are running into the underflow problems in a particular application?

> For cython, is it going to defeat the purpose of cython if I call scipy's logsumexp?

Unfortunately yes - calling a scipy function requires Cython to convert the memoryview back to a numpy array, and that overhead kills the benefits of using Cython.

> Seems like the only awkward part is adding a cython logsumexp function and getting it to work with the fancy templating.

I haven't thought about this much, but I don't think this would be a big deal. Since it's compiled to C, loops are no problem and log sum exp can be done just with a loop and the maximum element trick to protect against overflow / underflow.

The templating is really not too tough, but we do need to support the double complex case for computing derivatives via complex step differentiation.

**Comment by ChadFulton** [comment]
> so I guess the task now falls on me. :)

By the way, this would be great! I'm happy to help out.

**Comment by wmay** [comment]
I have a data set where the regimes are painfully obvious and switching is infrequent. I get some zero values from the filter function and then the kim smoother breaks due to division by zero.

**Comment by wmay** [comment]
I saw that you fixed the division by zero error in the new Kim smoother cython code. But I still get nan's when I pass zeros to the kim smoother.

I think the issue is in the next block of code, where infinite is potentially being multiplied by zero (pretty likely in my case). [Here](https://github.com/statsmodels/statsmodels/blob/master/statsmodels/tsa/regime_switching/_kim_smoother.pyx.in#L105). That turns into nan and then the nan's propagate from there.

I was hoping there'd be an easy way to patch the code without needing to convert everything to logs ... but I can't fix zero times infinite.

**Comment by rdturnermtl** [comment]
In general, I think going to log sum exp is the only way we can make this robust against corner cases.

**Comment by wmay** [comment]
To update, I did get a log version of this working, which turned my infinite/0 problem into an (infinite - infinite) problem. It's not obvious to me that underflow is distorting the results, so as far as I'm aware dealing with that corner case is all that's needed.

I ended up having to write custom code for my model anyway, so I'm not planning to push the log version (which is pretty disorganized) back to statsmodels. Though I'll be happy to post it if someone else wants to have a go at it.
